vyper/codegen/function_definitions/common.py

```diff
@@ -1,16 +1,37 @@
+from collections import defaultdict
 from dataclasses import dataclass
 from functools import cached_property
-from typing import Optional
+from typing import TYPE_CHECKING, Optional
 
+import vyper.ast as vy_ast
 from vyper.codegen.context import Constancy, Context
 from vyper.codegen.ir_node import IRnode
 from vyper.codegen.memory_allocator import MemoryAllocator
 from vyper.evm.opcodes import version_check
+from vyper.exceptions import CompilerPanic
 from vyper.semantics.types import VyperType
 from vyper.semantics.types.function import ContractFunctionT, StateMutability
 from vyper.semantics.types.module import ModuleT
 from vyper.utils import MemoryPositions
 
+if TYPE_CHECKING:
+    from vyper.semantics.analysis.base import VarInfo
+
+
+def analyse_last_use(fn_ast: vy_ast.FunctionDef):
+    counts: dict[VarInfo, int] = defaultdict(lambda: 0)
+    for stmt in fn_ast.body:
+        for expr in stmt.get_descendants(vy_ast.ExprNode):
+            info = expr._expr_info
+            if info is None:
+                continue
+            for r in info._reads:
```

Comment:
Not asking for a change, but I found it a bit confusing that _reads also include writes.